### DISCUSSION TITLE
Corfu not quitting (popup gets stuck) when using backward-char when editing Common Lisp

### DISCUSSION THREAD

**Discussion by skyler544** [minad/corfu#376] [Q&A]
First of all, thanks for this great package!

Here's the details for the simplest config that I could use to reproduce the bug:
`GNU Emacs 29.1 (build 2, x86_64-unknown-linux-gnu, GTK+ Version 3.24.38, cairo version 1.16.0) of 2023-08-03`

```lisp
;; init.el
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(use-package use-package
  :config
  (setq-default use-package-always-ensure t))

(use-package corfu
  :init
  (global-corfu-mode)
  :custom
  (corfu-auto t)
  (corfu-auto-prefix 2)
  (corfu-auto-delay 0)
  (corfu-preselect 'prompt)
  :bind
  (:map corfu-map
        ("TAB" . corfu-next)
        ([tab] . corfu-next)
        ("S-TAB" . corfu-previous)
        ([backtab] . corfu-previous)))

(use-package sly
  ;; Requires the `sbcl' binary to be available in your `exec-path'.
  :config (setq inferior-lisp-program "sbcl"))
```

This issue only happens in very specific circumstances and there's only one way I've been able to consistently reproduce it:

1. Open a Common Lisp buffer
2. Run `M-x sly` so that completions become available in the buffer
3. Type in `(lambda (elt|` with `|` signifying point. Corfu should have appeared for `lambda`, went away when you entered the space, then appeared again when you entered `elt`.
4. Type `C-b` or otherwise run `backward-char` at least four times so that you have this: `(lambda| (elt`
5. Observe that Corfu stays open and will remain open even if you enter a space (which with this config would normally force it to close) or move back to `(lambda (elt |` and start typing in something else like `foo`. Closing it seems to require a `C-g` at this point, and if you try to complete with the stuck popup you get `Corfu completion error: Wrong type argument: stringp, nil`.

Since this only happens in very specific circumstances and doesn't trigger with every half-baked thing you type in like in this contrived example, I can't really give much more information about what causes it (it occurs maybe once or twice an hour when editing Common Lisp normally, so isn't that frequent either). I was not able to reproduce this behavior in other situations like when editing C code with `eglot` and `clangd` providing the `capf`. It's not really critical either; you can just `C-g` the stuck popup away. However, you probably have a clearer idea of what could be causing it, so I thought I would submit this report anyways. 

Thanks again!

**Comment by minad** [comment]
Hi, I am not aware of this particular problem, but I recall another possibly related report, see #285.

**Reply by skyler544** [reply]
Seems related for sure; I thought it must have something to do specifically with the interaction between Corfu and Sly. 

My debugging ability is not up to the level of that reporter; I'll keep this in mind and see if I can investigate any further.

**Comment by skyler544** [comment]
Here's a workaround that prevents the most common case; this way you can still have the auto pop-up but for completion with less than four characters you'll need to activate corfu explicitly.

```lisp
(add-hook
 'lisp-mode-hook
 #'(lambda () (setq-local corfu-auto-prefix 4)))
```
